src/webnn/native/dml/GraphDML.cpp

```diff
@@ -1675,7 +1700,289 @@ namespace webnn::native ::dml {
+        DAWN_ASSERT(mGraphEdgesMap.find(inputsOperand[1].Get()) != mGraphEdgesMap.end());
+        DAWN_ASSERT(mGraphEdgesMap.find(inputsOperand[2].Get()) != mGraphEdgesMap.end());
+        std::vector<std::shared_ptr<EdgeInfoBase>> inputEdges;
+
+        // Input: 4D tensor with the Sizes of { 1, seq_length, batch_size, input_size }.
+        // Need to reshape input from WebNN 3-D to DML 4-D.
+        auto inputEdge = mGraphEdgesMap[inputsOperand[0].Get()];
+        auto webnnInputDims = ConvertDimensions(inputsOperand[0].Get()->Shape());
+        std::vector<UINT> inputDims = {1, webnnInputDims[0], webnnInputDims[1], webnnInputDims[2]};
+        std::shared_ptr<DmlTensorDesc> inputDmlTensorDesc(new DmlTensorDesc);
+        if (!CreateDmlTensorDesc(mDmlTensorsDesc, inputDmlTensorDesc, &inputEdge->outputTensorDESC,
+                                 inputDims)) {
+            return DAWN_INTERNAL_ERROR("Failed to create DML tensor description.");
+        }
+        DML_TENSOR_DESC inputTensorDesc = {DML_TENSOR_TYPE_BUFFER, &inputDmlTensorDesc->bufferDesc};
+        inputEdges.push_back(inputEdge);
+
+        // Weight: 4D tensor with the Sizes of { 1, num_directions, 3 * hidden_size, input_size }.
+        // Need to reshape weight from WebNN 3-D to DML 4-D.
+        // The TENSOR_FLAGS of weight, bias and hiddenInit in gru must be DML_TENSOR_FLAG_NONE.
+        ComPtr<IDMLOperator> dmlOperator;
+        auto constantWeightEdge = mGraphEdgesMap[inputsOperand[1].Get()];
+        auto webnnWeightDims = ConvertDimensions(inputsOperand[1].Get()->Shape());
+        std::vector<UINT> weightDims = {1, webnnWeightDims[0], webnnWeightDims[1],
+                                        webnnWeightDims[2]};
+        std::shared_ptr<DmlTensorDesc> constantWeightDmlTensorDesc(new DmlTensorDesc);
+        if (!CreateDmlTensorDesc(mDmlTensorsDesc, constantWeightDmlTensorDesc,
+                                 &constantWeightEdge->outputTensorDESC, weightDims)) {
+            return DAWN_INTERNAL_ERROR("Failed to create DML tensor description.");
+        }
+        // Workaround: append identity to convert constant input tensor with
+        // DML_TENSOR_FLAG_OWNED_BY_DML falg to input tenor with DML_TENSOR_FLAG_NONE flag.
+        DML_TENSOR_DESC constantWeightTensorDesc = {DML_TENSOR_TYPE_BUFFER,
+                                                    &constantWeightDmlTensorDesc->bufferDesc};
+        std::shared_ptr<DmlTensorDesc> weightDmlTensorDesc(new DmlTensorDesc);
+        if (!CreateDmlTensorDesc(mDmlTensorsDesc, weightDmlTensorDesc, &constantWeightTensorDesc,
+                                 {}, {}, true)) {
+            return DAWN_INTERNAL_ERROR("Failed to create DML tensor description.");
+        }
+        DML_TENSOR_DESC weightTensorDesc = {DML_TENSOR_TYPE_BUFFER,
+                                            &weightDmlTensorDesc->bufferDesc};
+        { APPEND_IDENTITY(constantWeightTensorDesc, weightTensorDesc, dmlOperator); }
```

Comment:
Can we make it as a function not a macro?

Reply:
Done, write it as a function of Graph class. PTAL, thanks!